### DISCUSSION TITLE
Flex Labels showing ellipsis on Mac

### DISCUSSION THREAD

**Discussion by markdemich** [airsdk/Adobe-Runtime-Support#2645] [Q&A]
I'm in the process of porting a air/flex application to use the Harman runtime while also moving away from Flashbuilder.

I have things pretty much working but I just realized when i run my app on the Mac, a lot of my Flex labels that should be autosized are showing ellipsis.  After doing some debugging, it seems the ones that are affected have internal textField width that is not an integer.  The ones I've seen end in .5.  I debugged against the Flex source and discovered that when the underlying TextField is parented and rendered it has a textWidth of 26.5.  However, there is a Flex object UITextFormat that is used to determine the width of text by creating a TextField but not parenting it and the using getLineMetrics(0) to return the size.  This size is coming up as 26 on my Mac.  The result is that it makes the logic think the Flex Label can't fit and it draws it with ellipsis.

Could this be something wrong with the version of Air I'm using.  It seems this is an issue with the runtime.  I'm using version 33.1.195, but previously I was one of the 50.x.x.x version (don't remember which since it was a mistake).  I want to stay as close to 32 as possible for now since fear moving to far away from that will cause other issues.

If anyone has any ideas of insight I would appreciate it. Been looking at this all day and I'm running out of ideas.

**Comment by ajwfrost** [comment]
Hi

I'd be interested to know whether this is only happening on Mac, or whether the same problem shows up on Windows too? But I think you've found an interesting point, that the textWidth of a TextField may differ depending on whether it's rendered into the display list or not... Ideally it should be the same, but there are sometimes some restrictions in that.. Plus it may be that there are differences in the fonts used or settings around kerning, borders etc when the text is rendered off-screen vs as part of the display list.

Having just tried a bunch of different text fields with random characters, interestingly the numbers always match perfectly on Windows .. but on macOS they often have a .5 discrepancy as you'd noticed. Only ever with the 'on-display' text field being at a something.5 width.

I was curious so checked this out with AIR 32 as well, same thing happens with that. So then I'm wondering, if you've not seen this before, whether there's something in the compiler settings that's changing it (i.e. an app built with the 'old' Flex compiler, vs an app built with the 'new' AS3 compiler).. but I can't really see how!

We can treat this as a bug, but if you're trying not to update your runtime too much, then you may need to update the Flex code to look at the values (or just always add 0.5 to any value given by the UITextFormat off-screen text width calculation)

thanks

**Comment by markdemich** [comment]
Thanks for looking at this.  Yes, it's only  happening on the Mac.  Also, I
have an old Mac (Catalina).   Not sure if that matters.

Also, it was working fine with runtime 32.0.0.170 (I believe this was the
adobe one).

It could be a setting. In fact, I spent most of yesterday comparing
settings but could not find any differences besides how I compile and
build.  I did need to re-organize that part since I'm not in Flash
Builder anymore so that is suspect. You just gave me an idea though. I can
point my new VSCode setup with the as3mxl extension at the old runtime and
rebuild and see if the problem persists or not.  I'll let you know how I
make out.

For now, I edited the logic in flex to add 1.0 (but I think I'll try 0.5
like you suggest) to that one calculation that is not returning the 0.5.
This is just surfacing as a problem with the ellipsis calculation so I can
isolate my hack for that, and I'll probably put some kind of "isMac"
conditional in there as well for good measure to avoid potential side
effects in windows.

Thanks for being so responsive to my questions. I appreciate it.

**Reply by markdemich** [reply]
I downloaded the the 32 version of the sdk with compiler from adobe's site here - https://helpx.adobe.com/air/kb/archived-air-sdk-version.html#main_Archived_Adobe_AIR_downloads_. I then tried to overlay that on top of my flex 4.1 sdk, but I wound up with lots of errors attempting to compile.  I'm not sure this experiment is worth the effort since I found a work around.  Here are some of the error was getting.  Maybe something will jump out to you.

May 08, 2023 9:45:50 AM com.google.common.base.internal.Finalizer getInheritableThreadLocalsField
INFO: Couldn't access Thread.inheritableThreadLocals. Reference finalizer threads will inherit thread local values.

Error: /Users/..../......mxml is of an unsupported type: mxml

command line
Warning: 'compiler.locale' is not fully supported.

command line
Warning: 'compiler.theme' is not fully supported.

command line
Warning: 'compiler.use-resource-bundle-metadata' is no longer supported and will have no effect.

command line
Warning: 'compiler.mxml.compatibility-version' is not fully supported.

command line
Warning: 'compiler.fonts.managers' is not fully supported.

command line
Warning: 'compiler.incremental' is no longer supported and will have no effect.

command line
Warning: 'compiler.accessible' is not fully supported.

**Reply by joshtynjala** [reply]
> I downloaded the the 32 version of the sdk with compiler from adobe's site here
> I then tried to overlay that on top of my flex 4.1 sdk

You can't mix the compilers together int the same SdK. If you want to use the Flex SDK, you always need to download the version of the AIR SDK without a compiler.

**Reply by markdemich** [reply]
Oh, interesting.  I'll try that.  Thanks.

**Reply by markdemich** [reply]
I was able to compile and run against version 32 of the runtime and the problem persists.  So this looks like it's somehow related to how I'm compiling and packaging things.  Or maybe it's about running in Catalina?  For now, I need to move on so my work-around of changing UITextField.as makes things functional so I'll have to live with this until it presents itself as blocking issue later.

Thanks for your help.

**Reply by ajwfrost** [reply]
It's more about the runtime I think .. it looks like when you render on the display list on a Mac, it will use an antialiasing approach and will render onto a higher-resolution target; off-screen it's just rendering on a per-pixel basis. So this means you can sometimes get half-pixel width values from the text width on a Mac.

A simple solution for us here is to just apply the same anti-aliasing on the off-screen text measurement for Mac .. so we can do that as a bug fix and from checking it here, the two values are now always the same, even if they're half-pixel values.

thanks
